### PR TITLE
Add docker networks to the Overlay Topology

### LINKED ISSUE #1563: Missing edges when using Docker's IPAM driver
Scope is not displaying some connections while we are running continuous load through our application.

Specifically, the front-end service is only shown as having one outbound connection, to the catalogue service. In fact it is communicating with the login, orders, accounts, and carts services as well. I have verified this via our log files.

Please ping me if you need any additional information.

[weave_report_and_svg.zip](https://github.com/weaveworks/scope/files/304731/weave_report_and_svg.zip)

### PR DESCRIPTION
**Pull request by 2opremio**
Fixes #1563

### CODE REVIEW

File: probe/docker/registry.go [hunk 1]
```diff
@@ -249,7 +262,26 @@ func (r *registry) updateImages() error {
...
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	// reset
+	r.networks = r.networks[:0]
+	for i := range networks {
+		r.networks = append(r.networks, &networks[i])
+	}
```
Comment (tomwilkie) [inline probe/docker/registry.go]: Why not just r.networks = networks?

File: probe/docker/reporter.go [hunk 2]
```diff
@@ -241,6 +244,21 @@ func (r *Reporter) containerImageTopology() report.Topology {
...
+			for _, config := range network.IPAM.Config {
+				localSubnets = append(localSubnets, config.Subnet)
+			}
+		}
+	})
+	peerID := OverlayPeerPrefix + r.hostID
+	node := report.MakeNode(report.MakeOverlayNodeID(peerID)).WithSets(
+		report.MakeSets().Add(host.LocalNetworks, report.MakeStringSet(localSubnets...)))
+	return report.MakeTopology().AddNode(node)
+}
```
Comment (tomwilkie) [inline probe/docker/reporter.go]: Whats this supposed to achieve?

File: probe/docker/registry.go (hunk 1 above)
Comment (2opremio) [inline probe/docker/registry.go]: You are right

**Comment by tomwilkie** [comment]
I don't understand how this fixed that issue.

**Comment by tomwilkie** [comment]
Okay explained f2f.  Let me think for 20mins.

**Comment by 2opremio** [comment]
As part of the review feedback commit I have changed the registry to store (and walk) images instead of image pointers (we are not modifying them so AFAIU there was no reason to use pointers).

**Comment by tomwilkie** [comment]
@idcrosby @2opremio as discussed, need a reproduction for this pls.

**Comment by tomwilkie** [comment]
Try to reproduce the issue (without this PR):

```
$ docker network create foo
$ docker run --net=foo -d --name nginx nginx
$ docker run --net=foo -d --name client alpine /bin/sh -c "while true; do \
wget http://nginx:80/ -O - >/dev/null || true; \
sleep 1; \
done"
```

But the edges appeared.

**Comment by tomwilkie** [comment]
Okay if you use the weavemesh driver I can indeed reproduce:

```
$ weave launch
$ docker network create --driver=weavemesh foo
$ docker run --net=foo -d --name nginx nginx
$ docker run --net=foo -d --name client alpine /bin/sh -c "while true; do \
wget http://nginx:80/ -O - >/dev/null || true; \
sleep 1; \
done"
```

**Comment by tomwilkie** [comment]
/cc @rade 

screen shot 2016-06-15 at 16 48 57

**Comment by tomwilkie** [comment]
back to fons as didn't fix the above repro.

**Comment by 2opremio** [comment]
Meh, the problem is that the go client for some reason is not filling in the `IPAMConfig` field:

```
docker.Network{
        Name:   "foo",
        ID:     "fdbd13e9e24280f38e5a3810b8b55352c1d1cd099ad34a143414fe3e39a01501",
        Scope:  "local",
        Driver: "weavemesh",
        IPAM:   docker.IPAMOptions{Driver: "default", Config: []docker.IPAMConfig(nil)},
        Containers: map[string]docker.Endpoint{
                "d42139eae4697224b5deefbdaaefb927345241b62b55ca7f294e81cdc046bd7a": docker.Endpoint{
                        Name:        "nginx",
                        ID:          "30dfb5ec8abf5c882701b154e9e87439c2f2286154f72a913a34b25c3f58c569",
                        MacAddress:  "",
                        IPv4Address: "172.18.0.3/16",
                        IPv6Address: ""},
                "ec3ab14f5543c0567c84030c2585a0786aa0ad140f99dc7db3fc16bde3149b29": docker.Endpoint{
                        Name:        "client",
                        ID:          "50004dbcaca6bdca3fda7dcc24526fa61be7e23aac02a22dcc7f96df56ca0ced",
                        MacAddress:  "",
                        IPv4Address: "172.18.0.2/16",
                        IPv6Address: ""}},
        Options:  map[string]string{},
        Internal: false,
}
```

```
$ docker network inspect foo -o json                                                                                                                                
[
    {
        "Name": "foo",
        "Id": "fdbd13e9e24280f38e5a3810b8b55352c1d1cd099ad34a143414fe3e39a01501",
        "Scope": "local",
        "Driver": "weavemesh",
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.18.0.0/16",
                    "Gateway": "172.18.0.1/16"
                }
            ]
        },
        "Containers": {
            "d42139eae4697224b5deefbdaaefb927345241b62b55ca7f294e81cdc046bd7a": {
                "Name": "nginx",
                "EndpointID": "30dfb5ec8abf5c882701b154e9e87439c2f2286154f72a913a34b25c3f58c569",
                "MacAddress": "",
                "IPv4Address": "172.18.0.3/16",
                "IPv6Address": ""
            },
            "ec3ab14f5543c0567c84030c2585a0786aa0ad140f99dc7db3fc16bde3149b29": {
                "Name": "client",
                "EndpointID": "50004dbcaca6bdca3fda7dcc24526fa61be7e23aac02a22dcc7f96df56ca0ced",
                "MacAddress": "",
                "IPv4Address": "172.18.0.2/16",
                "IPv6Address": ""
            }
        },
        "Options": {}
    }
]
```

My bad: https://twitter.com/ThePracticalDev/status/687672086152753152

**Comment by 2opremio** [comment]
Updating the docker client made it work.

**Comment by tomwilkie** [comment]
LGTM

**Pull request merged**